Fix AsyncPipe types for RxJS 6 and 7

benlesh opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] ~~Tests for the changes have been added (for bug fixes / features)~~ this is a non-breaking change and adding a test would require adding a development dependency to this project of `rxjs@next`, which seems like overkill. However this will get exercised extremely well in `google3`.
- [x] ~~Docs have been added / updated (for bug fixes / features)~~ unnecessary


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [x] Other... Please describe:

This resolves an issue reported to me by @leggechr and escalated to @alxhub. The issue is discussed here as well: https://github.com/microsoft/TypeScript/issues/43643




## What is the current behavior?
Currently, when upgrading to RxJS 7, in many cases Angular templates are unable to infer the proper value type from RxJS observables. This is because of the TypeScript issue/quirk pointed out above.


## What is the new behavior?
The value types will be inferred properly in the Angular template.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

This is affecting targets within `google3` and I recommend you work with @leggechr and escalate this as a Googler issue.

martinmcwhorter commented on packages/common/src/pipes/async_pipe.ts:
```diff
@@ -84,7 +84,7 @@ export class AsyncPipe implements OnDestroy, PipeTransform {
 
   private _subscription: Unsubscribable|Promise<any>|null = null;
-  private _obj: Subscribable<any>|Promise<any>|EventEmitter<any>|null = null;
+  private _obj:|Subscribable<any>|Promise<any>|EventEmitter<any>|null = null;
```
Leading |
Did you mean to add Observable?

alxhub replied:
Yeah, this change seems to have been corrupted before it was pushed.

benlesh replied:
Weird, not sure what happened here.

alxhub commented on packages/common/src/pipes/async_pipe.ts:
```diff
@@ -39,7 +39,7 @@ class SubscribableStrategy implements SubscriptionStrategy {
 class PromiseStrategy implements SubscriptionStrategy {
   createSubscription(async: Promise<any>, updateLatestValue: (v: any) => any): Promise<any> {
-    return async.then(updateLatestValue, e => {
+    return async.then(updateLatestValue, (e) => {
```
Spurious change?

benlesh replied:
formatter, most likely.

benlesh replied:
Okay, reverted everything, disabled my local format on save option, and tried again. Hopefully this resolves the issue.

benlesh committed:
fix(core): AsyncPipe now compatible with RxJS 7

Adds a fix to make sure that RxJS v7 Observable is compatible with AsyncPipe. This is a typings-only change.
For more information see: https://github.com/microsoft/TypeScript/issues/43643

e-oz commented:
Sorry if it's a noobish request, but could one of you help bots to complete the checks? Please.
I'd like to try RxJS 7 with the Angular :)

petebacondarwin reviewed (approved):
LOL! We only just recently switched this from Observable to Subscribable!

I am not sure where this puts us in terms of breaking changes... it is possible that, since we made the change, someone has implemented this pipe with a `Subscribable` object rather than an `Observable`, which would no longer work. But I suspect this is very unlikely.

Approving generally. This should trigger a public-api review round.

AndrewKushnir commented:
[Presubmit](http://test/OCL:371516472:BASE:371516478:1619905190091:368f21cf).

alxhub commented:
> I am not sure where this puts us in terms of breaking changes... it is possible that, since we made the change, someone has implemented this pipe with a `Subscribable` object rather than an `Observable`, which would no longer work. But I suspect this is very unlikely.

Since this change only adds `Observable<T>` to the signature and doesn't remove `Subscribable<T>`, such a usage should still work.

alxhub reviewed (approved):
Reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
I _Observe_ that these changes look good.

reviewed-for: public-api

AndrewKushnir commented:
FYI, presubmit looks good, the changes were approved (and the changes look backwards-compatible), so I'm adding the "merge" label. Thank you.
